pkg/services/grpcserver/query.go

```diff
@@ -0,0 +1,174 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/genproto/pluginv2"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/expr"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/adapters"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/org"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/tsdb/grafanads"
+)
+
+type QueryDataService struct {
+	cfg    *setting.Cfg
+	server *queryDataServer
+}
+
+func ProvideQueryDataService(cfg *setting.Cfg, grpcServerProvider Provider, pluginsClient plugins.Client, dataSourceCache datasources.CacheService, dataSourceService datasources.DataSourceService) (*QueryDataService, error) {
+	server := &queryDataServer{
+		logger:            log.New("grpc-server-query-data"),
+		pluginClient:      pluginsClient,
+		dataSourceCache:   dataSourceCache,
+		dataSourceService: dataSourceService,
+	}
+	pluginv2.RegisterDataServer(grpcServerProvider.GetServer(), server)
+	return &QueryDataService{
+		cfg:    cfg,
+		server: server,
+	}, nil
+}
+
+type queryDataServer struct {
+	logger            log.Logger
+	pluginClient      plugins.Client
+	dataSourceCache   datasources.CacheService
+	dataSourceService datasources.DataSourceService
+}
+
+// QueryData handles multiple queries run against a single datasource
+func (s *queryDataServer) QueryData(ctx context.Context, in *pluginv2.QueryDataRequest) (*pluginv2.QueryDataResponse, error) {
+	req := backend.FromProto().QueryDataRequest(in)
+
+	// This allows the plugin to execute queries 'on behalf of' any user
+	// Given that the connection requires an admin token to run, this is OK
+	// However, we need to implement a more robust auth/token passing system
+	user := signedInUserFromBackendUser(req.PluginContext.User)
+	user.OrgID = req.PluginContext.OrgID
+
+	requests, err := s.buildRequests(ctx, user, req)
+	if err != nil {
+		return nil, err
+	}
+
+	response := backend.NewQueryDataResponse()
+
+	for _, req := range requests {
+		res, err := s.pluginClient.QueryData(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		for refID, res := range res.Responses {
+			response.Responses[refID] = res
+		}
+	}
+
+	return backend.ToProto().QueryDataResponse(response)
+}
+
+// TODO: can this be shared with the query service?
+func (s *queryDataServer) getDataSourceFromQuery(ctx context.Context, signedInUser *user.SignedInUser, skipCache bool, query *simplejson.Json) (*datasources.DataSource, error) {
+	uid := query.Get("datasource").Get("uid").MustString()
```

Comment:
NOTE: although this is not in the official query model, it is added to every query on the frontend

Reply:
this was originally implemented this way because the plugin that is sending a query doesn't have access to the instance settings for another datasource, so the plugin context couldn't be used. a nice side effect is that it supports mixing queries to different datasources in a single `QueryDataRequest`